### PR TITLE
Fix Random.Next(int, int) on 64-bit when max-min overflows int.MaxValue

### LINKED ISSUE #50911: Random.Next(int,int) can return out-of-bounds result in .NET 6 Preview 2
### Description

[My UnitTests](https://github.com/koszeggy/KGySoft.CoreLibraries/blob/54d58f3eef59ff97059ee9752a53be0a12348c60/KGySoft.CoreLibraries.UnitTest/UnitTests/CoreLibraries/Extensions/RandomExtensionsTest.cs#L131) revealed the following bug when targeting .NET 6 Preview 2:

The [`Random(int,int)`](https://github.com/dotnet/runtime/blob/02e951f1e42ae2e9759bd5a3267f5bc2407e5ab6/src/libraries/System.Private.CoreLib/src/System/Random.cs#L82) method can return invalid results.

The following example has about 50% chance to return a huge negative value:

```cs
int i = new Random().Next(-1, Int32.MaxValue);
```

### Configuration

Reproducible with .NET 6 Preview 2

### Other information

The bug is reproducible only if there is no seed specified so the new `XoshiroImpl` is created internally.

### PR DESCRIPTION
**Pull request by stephentoub**
The fix is a missing cast.  Everything else is asserts and test changes.

Fixes https://github.com/dotnet/runtime/issues/50911

### CODE REVIEW

**Review by koszeggy — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Random.cs [hunk 1]
```diff
@@ -155,14 +183,36 @@ public virtual void NextBytes(byte[] buffer)
...
+            return result;
+        }
 
         private static void ThrowMaxValueMustBeNonNegative() =>
             throw new ArgumentOutOfRangeException("maxValue", SR.Format(SR.ArgumentOutOfRange_NeedNonNegNum, "maxValue"));
 
         private static void ThrowMinMaxValueSwapped() =>
             throw new ArgumentOutOfRangeException("minValue", SR.Format(SR.Argument_MinMaxValue, "minValue", "maxValue"));
 
+        [Conditional("DEBUG")]
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/Random.cs:199]: One would hope the jit would see that all the code inside this method is debug only and this annotation would be superfluous

File: src/libraries/System.Runtime.Extensions/tests/System/Random.cs [hunk 2]
```diff
@@ -178,6 +202,30 @@ public void Next_LongLong_AllValuesWithinSmallRangeHit(bool derived, bool seeded
...
+                (long.MinValue, 0L),
+            }
+            select new object[] { derived, seeded, pair.min, pair.max };
+
+        [Theory]
+        [MemberData(nameof(Next_LongLong_Next_IntInt_AllValuesAreWithinRange_MemberData))]
+        public void Next_LongLong_Next_IntInt_AllValuesAreWithinRange(bool derived, bool seeded, long min, long max)
+        {
+            Random r = Create(derived, seeded);
+            for (int i = 0; i < 100; i++)
```
Comment (danmoseley) [inline src/libraries/System.Runtime.Extensions/tests/System/Random.cs:223]: If we had a bug, would this test realistically discover it? 
I assume that 100 could be quite a bit larger without significant time cost.

File: src/libraries/System.Runtime.Extensions/tests/System/Random.cs [hunk 3]
```diff
@@ -130,6 +130,30 @@ public void Next_IntInt_AllValuesWithinSmallRangeHit(bool derived, bool seeded)
...
+        public static IEnumerable<object[]> Next_IntInt_Next_IntInt_AllValuesAreWithinRange_MemberData() =>
+            from derived in new[] { false, true }
+            from seeded in new[] { false, true }
+            from (int min, int max) pair in new[]
+            {
+                (1, 2),
+                (-10, -3),
+                (0, int.MaxValue),
+                (-1, int.MaxValue),
+                (int.MinValue, 0),
```
Comment (danmoseley) [inline src/libraries/System.Runtime.Extensions/tests/System/Random.cs:142]: Pass in both minvalue and maxvalue together? And below

**Review by danmoseley — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Random.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Random.cs:199]: With the attribute it's removed by the C# compiler long before it gets to the jit :)

File: src/libraries/System.Runtime.Extensions/tests/System/Random.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Runtime.Extensions/tests/System/Random.cs:223]: We did have a bug,  and it typically only took two iterations to flag it.  I could make it larger if you like,  though.

Comment (danmoseley) [reply src/libraries/System.Runtime.Extensions/tests/System/Random.cs:223]: I don't feel strongly

**Pull request merged**